src/xtea.cpp

```diff
@@ -30,111 +30,46 @@ namespace {
-    for (auto i = 0u, j = 0u; i < BLOCK_SIZE; i += 1u, j += 8u) {
-        left[i] = data[j] | data[j+1] << 8u | data[j+2] << 16u | data[j+3] << 24u;
-        right[i] = data[j+4] | data[j+5] << 8u | data[j+6] << 16u | data[j+7] << 24u;
-    }
-
-    uint32_t sum = delta << 5;
-    for (auto i = 0u; i < 32; ++i) {
-        for (auto j = 0u; j < BLOCK_SIZE; ++j) {
-            right[j] -= (((left[j] << 4) ^ (left[j] >> 5)) + left[j]) ^ (sum + k[(sum >> 11) & 3]);
-        }
-        sum -= delta;
-        for (auto j = 0u; j < BLOCK_SIZE; ++j) {
-            left[j] -= (((right[j] << 4) ^ (right[j] >> 5)) + right[j]) ^ (sum + k[(sum) & 3]);
-        }
-    }
-
-    for (auto i = 0u, j = 0u; i < BLOCK_SIZE; i += 1u, j += 8u) {
-        data[j] = static_cast<uint8_t>(left[i]);
-        data[j+1] = static_cast<uint8_t>(left[i] >> 8u);
-        data[j+2] = static_cast<uint8_t>(left[i] >> 16u);
-        data[j+3] = static_cast<uint8_t>(left[i] >> 24u);
-        data[j+4] = static_cast<uint8_t>(right[i]);
-        data[j+5] = static_cast<uint8_t>(right[i] >> 8u);
-        data[j+6] = static_cast<uint8_t>(right[i] >> 16u);
-        data[j+7] = static_cast<uint8_t>(right[i] >> 24u);
-    }
+    for (auto i = 0u, sum = 0u, next_sum = sum + delta; i < 32u; ++i, sum = next_sum, next_sum += delta) {
+        apply_rounds(data, length, [&](uint32_t& left, uint32_t& right) {
+            left += ((right << 4 ^ right >> 5) + right) ^ (sum + k[sum & 3]);
+            right += ((left << 4 ^ left >> 5) + left) ^ (next_sum + k[(next_sum >> 11) & 3]);
+        });
+    };
 }
 
-constexpr auto InitialBlockSize =
-#if defined(__AVX512F__)
-      128u;
-#elif defined(__AVX__)
-      32u;
-#elif defined(__SSE__) || defined(__ARM_FEATURE_SIMD32)
-      8u;
-#elif defined(__x86_64__)
-      2u;
-#else
-      1u;
-#endif
-
-template<bool Encrypt, size_t BlockSize>
-struct XTEA {
-    static constexpr auto step = BlockSize * 8u;
-
-    void operator()(uint8_t* input, size_t length, const key& k) const {
-        const auto blocks = (length & ~(step - 1));
-        for (auto i = 0u; i < blocks; i += step) {
-            if (Encrypt) {
-                XTEA_encrypt<BlockSize>(input + i, k);
-            } else {
-                XTEA_decrypt<BlockSize>(input + i, k);
-            }
-        }
-        input += blocks;
-        length -= blocks;
-
-        if (BlockSize != 1) {
-            XTEA<Encrypt, (BlockSize + 1u) / 2u>()(input, length, k);
-        }
-    }
-};
-
-constexpr auto encrypt_v = XTEA<true, InitialBlockSize>();
-constexpr auto decrypt_v = XTEA<false, InitialBlockSize>();
-
-} // anonymous namespace
-
-void encrypt(uint8_t* data, size_t length, const key& k) { encrypt_v(data, length, k); }
-void decrypt(uint8_t* data, size_t length, const key& k) { decrypt_v(data, length, k); }
+void decrypt(uint8_t* data, size_t length, const key& k)
+{
+    for (auto i = 0u, sum = delta * 32, next_sum = sum - delta; i < 32u; ++i, sum = next_sum, next_sum -= delta) {
```

Comment:
Why not use `uint64_t` instead of `auto` ? this removes the `C26450` warning.

Reply:
Compiling in 32 bits there is this warning in both visual studio 2015 and 2019: "warning C4307: '*': integral constant overflow". long long fix the warning.

Reply:
@MillhioreBT it should actually infer `uint32_t`. Using `auto` is personal taste, I will change it due to the warning.

Reply:
@CristoferMartins how do I tell the VS compiler the overflow it is intended?

Reply:
Did it fix for you both?

Reply:
> @MillhioreBT it should actually infer `uint32_t`. Using `auto` is personal taste, I will change it due to the warning.

@ranisalt yeah it is correct the use of `uint32_t`